price-discovery/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+    ) -> SCResult<()> {
+        require!(
+            self.blockchain().is_smart_contract(&dex_sc_address),
+            "Invalid DEX SC address"
+        );
+        require!(
+            launched_token_id.is_valid_esdt_identifier(),
+            "Invalid launched token ID"
+        );
+        require!(
+            accepted_token_id.is_egld() || accepted_token_id.is_valid_esdt_identifier(),
+            "Invalid payment token ID"
+        );
+
+        let current_epoch = self.blockchain().get_block_epoch();
+        require!(
+            current_epoch < start_epoch,
+            "Start epoch cannot be in the past"
+        );
+        require!(current_epoch < end_epoch, "End epoch cannot be in the past");
+        require!(
+            start_epoch < end_epoch,
+            "Start epoch must be before end epoch"
+        );
+
+        self.dex_sc_address().set(&dex_sc_address);
+        self.launched_token_id().set(&launched_token_id);
+        self.accepted_token_id().set(&accepted_token_id);
+        self.start_epoch().set(&start_epoch);
+        self.end_epoch().set(&end_epoch);
+
+        Ok(())
+    }
+
+    #[only_owner]
+    #[payable("*")]
+    #[endpoint(depositInitialTokens)]
+    fn deposit_initial_tokens(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let start_epoch = self.start_epoch().get();
+        require!(
+            current_epoch < start_epoch,
+            "May only deposit before start epoch"
+        );
+
+        let launched_token_id = self.launched_token_id().get();
+        require!(payment_token == launched_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, LAUNCHED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn deposit(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let accepted_token_id = self.accepted_token_id().get();
+        require!(payment_token == accepted_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let caller = self.blockchain().get_caller();
+        self.mint_and_send_redeem_token(&caller, ACCEPTED_TOKEN_REDEEM_NONCE, &payment_amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn withdraw(
+        &self,
+        #[payment_token] payment_token: TokenIdentifier,
+        #[payment_nonce] payment_nonce: u64,
+        #[payment_amount] payment_amount: BigUint,
+    ) -> SCResult<()> {
+        self.require_active()?;
+
+        let redeem_token_id = self.redeem_token_id().get();
+        require!(
+            payment_token == redeem_token_id && payment_nonce == ACCEPTED_TOKEN_REDEEM_NONCE,
```

Comment:
withdraw can be done for accepted_token and launchpad_token as well. Some users have tokenA, others have tokenB they can simply enter with either of them. As the price discovery goes on, some users might want to withdraw their position - because they do not like how the price will be once the LP is launch. They can withdraw tokenA or tokenB.

Reply:
Done.